core/src/main/scala/fs2/async/mutable/Signal.scala

```diff
@@ -62,60 +47,46 @@ object Signal {
-          val modify =
-            F.modify(ref){ case s@(a,current,q) =>
-              if (current != stamp) F.pure(s)
-              else F.pure((a,current,q :+ {(va:((Int,A))) => F.set(chref)(F.pure(va))}))
-            }
-
-          F.bind(modify){
-            case Change((current,a,_), _) =>
-              if (current != stamp) F.pure((current,a))
-              else F.get(chref)
-          }
-        }
-      }
-
-      def getStamped:F[(Int,A)] = F.map(F.get(ref)){case (v,a,_) => (v,a)}
-
-      new Signal[F,A] {
-        def refresh: F[Unit] = F.map(compareAndSet(a => Some(a)))(_ => ())
-        def set(a: A): F[Unit] = F.map(compareAndSet(_ => Some(a)))(_ => ())
-        def get: F[A] = F.map(F.get(ref))(_._2)
-        def compareAndSet(op: (A) => Option[A]): F[Option[A]] = {
-          val modify:F[Change[State[F,A]]] =
-            F.modify(ref) { case (v,a,q) => F.pure(op(a).fold((v,a,q)){ na => (v+1,na,SQueue.empty)}) }
-
-          F.bind(modify) {
-           case Change((oldVersion,_,queued),(newVersion,newA,_)) =>
-             if (oldVersion == newVersion) F.pure(None:Option[A])
-             else {
-               queued.foldLeft(F.pure(Option(newA))) {
-                 case (r,f) =>  F.bind(f(newVersion -> newA))(_ => r)
-               }
-             }
+  private class ID
+
+  def apply[F[_],A](initA: A)(implicit F: Async[F]): F[Signal[F,A]] =
+    F.map(F.refOf[(A, Long, LinkedMap[ID, Semaphore[F]])]((initA,0,LinkedMap.empty))) {
+    state => new Signal[F,A] {
+      def refresh: F[Unit] = F.map(possiblyModify(a => Some(a)))(_ => ())
+      def set(a: A): F[Unit] = F.map(possiblyModify(_ => Some(a)))(_ => ())
+      def get: F[A] = F.map(F.get(state))(_._1)
+      def possiblyModify(f: A => Option[A]): F[Option[A]] =
+        F.bind(F.modify(state) { state => f(state._1) match {
+          case None => state
+          case Some(a2) => (a2, state._2 + 1, state._3)
+        }}) { c =>
+          if (c.now._2 != c.previous._2) { // set was successful
+            F.map(F.traverse(c.now._3.values.toVector) { semaphore =>
+              F.bind(semaphore.tryDecrement) { _ => semaphore.increment }
+            }) { (vs: Vector[Unit]) => Some(c.now._1) }
           }
+          else F.pure(None)
         }
 
-        def changes: fs2.Stream[F, Unit] =
+        def changes: Stream[F, Unit] =
           discrete.map(_ => ())
 
-        def continuous: fs2.Stream[F, A] =
-          Stream.eval(get).append(continuous)
-
-        def discrete: fs2.Stream[F, A] = {
-          def go(stamp:Int):fs2.Stream[F, A] = {
-            Stream.eval(getChanged(stamp))
-            .flatMap { case (s,a) => Stream(a).append(go(s)) }
-          }
-          Stream.eval(getStamped).flatMap { case (s,a) => Stream(a).append(go(s))}
+        def continuous: Stream[F, A] =
+          Stream.repeatEval(get)
+
+        def discrete: Stream[F, A] = {
+          val s: F[(ID, Semaphore[F])] =
+            F.bind(Semaphore(0)) { s =>
```

Comment:
@pchlupacek If we want `discrete` to get the initial value also, we would just change this to `Semaphore(1)`.